clrs/_src/clrs_text/huggingface_generators.py

```diff
@@ -0,0 +1,122 @@
+"""Functions to allow for Huggingface Intergration"""
+from typing import List, Dict
+from clrs import build_sampler
+from clrs._src.clrs_text.clrs_utils import format_clrs_example
+import itertools
+
+
+def clrs_gen(algs: Dict[str, List[int]], num_samples: int, seed: int = 0, use_hints: bool = True, pretrain: bool = False):
+    """
+    Huggingface datasets.Dataset compatible generator function for creating a dataset of fixed size
+
+    Example usage:
+        import datasets
+        algs = {"insertion_sort": [16]}
+        ds = datasets.Dataset.from_generator(clrs_gen, gen_kwargs={"algs": algs, "num_samples": 100})
+
+    Huggingface reference:
+        https://huggingface.co/docs/datasets/v2.19.0/en/package_reference/main_classes#datasets.Dataset.from_generator
+
+    Args:
+        algs (Dict[str, List[int]]): keys = algorithm names [Must be same as in clrs.CLRS_30_ALGS_SETTINGS.keys()], values = list of lengths required for that algorithm
+        num_samples (int): The size of the output dataset
+        seed (int), default=0: The random seed for all of the generators
+        use_hints (bool), default=True: Whether to include hints in the questions
+        pretrain (bool), default=False: Whether to combine the question and answer into one field called text
+ 
+    Returns:
+        Sample question and answer as a dictionary
+    """
+    choices = {}
+
+    # make all of the possible generators
+    for alg in algs.keys():
+        for length in algs[alg]:
+            sampler, _ = build_sampler(
+                alg,
+                seed=seed,
+                num_samples=-1,
+                length=length,
+                track_max_steps=False,
+                use_padding=False,
+            )
+            choices[f"{alg}={length}"] = sampler
+
+    keys = itertools.cycle(list(choices.keys())) # make a cyclic list of all the generators
```

Comment:
Current approach seems to be quite complex and ineffective -- why don't we just randomly sample at each iteration instead of generation of keys loop and going through it.

While it can be complicated to do that from dict, we can store a list instead, which will have `(algo_name, length, sampler)` entries instead of `f"{algo_name}={length}": sampler` for each sampler.

So instead of `choices[f"{alg}={length}"] = sampler` you can  do this `choices.append(algo_name, length, sampler)` and then here do something like this:
```python
import random
...
random.seed(seed)  # outside of loop
...
algo_name, length, sampler = random.sample(clrs_samplers, 1)
sample = sampler.next(batch_size=1)
```